composer/datasets/in_context_learning_evaluation.py

```diff
@@ -0,0 +1,120 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+import textwrap
+
+import torch
+from datasets import load_dataset
+from torch.utils.data import DataLoader, Dataset
+from transformers import AutoTokenizer
+
+from composer.core import DataSpec
+from composer.utils import dist, ensure_tuple
+from composer.utils.file_helpers import get_file
+
+
+class InContextLearningLMTaskDataset(Dataset):
+
+    def __init__(
+        self,
+        dataset_uri: str,
+        tokenizer: AutoTokenizer,
+        max_seq_len: int,
+        eos_tok_id: int,
+        destination_path: str = 'icl_lm_task.json',
+    ):
+        get_file(dataset_uri, destination_path, overwrite=True)
+        dataset = load_dataset('json', data_files=destination_path, split='train', streaming=False)
+        self.encoded_dataset = list(
+            dataset.map(lambda examples: {
+                'continuation': tokenizer(examples['continuation']),
+                'context': tokenizer(examples['context']),
+            }))
+        self.tokenizer = tokenizer
+        self.max_seq_len = max_seq_len
+        self.eos_tok_id = eos_tok_id
+
+    def __getitem__(self, index):
+        return self.encoded_dataset[index]
+
+    def __len__(self):
+        return len(self.encoded_dataset)
+
+    def collate_fn(self, data):
+        inputs = []
+        continuation_indices = []
+        for data_pair in data:
+            context, continuation = data_pair['context'], data_pair['continuation']
+
+            context_enc = context['input_ids']
+            continuation_enc = continuation['input_ids']
+            continuation_span = torch.tensor(range(len(context_enc), len(context_enc) + len(continuation_enc)))
+
+            inp = torch.tensor(
+                (context_enc + continuation_enc)[-(self.max_seq_len + 1):],
+                dtype=torch.long,
+            )
+            (inplen,) = inp.shape
+
+            # pad length from seq to padding_length
+            inp = torch.cat(
+                [
+                    inp,  # [seq]
+                    torch.LongTensor((self.max_seq_len - inplen) * [self.eos_tok_id]),  # [padding_length - seq]
+                ],
+                dim=0,
+            )
+
+            inputs.append(inp)
+            continuation_indices.append(continuation_span)
+
+        return {
+            'input_ids': torch.stack(inputs),
+            'continuation_indices': continuation_indices,
+            'mode': 'lm_task',
+            'labels': torch.stack(inputs),
+            'eos_tok_id': self.eos_tok_id
+        }
+
+    def get_num_samples_in_batch(self, batch) -> int:
```

Comment:
What is this needed for? It looks like it is exactly the same as the default function: https://github.com/mosaicml/composer/blob/1a3f210e8a714a702e0f14409a93d4141f0d171c/composer/core/data_spec.py#L277

Reply:
The only difference is this little section ` if isinstance(t, torch.Tensor)` to ensure we don't try taking the shape of any of the non-tensor values in the batch dict